### PR TITLE
Xmc4800 automation board support and MEMC drivers for XMC4xxx

### PR DESCRIPTION
**Pull request by talih0**
- Includes MEMC drivers for XMC4xxx
- The dev kit has an SDRAM connected via MEMC.

### CODE REVIEW

File: dts/bindings/memory-controllers/infineon,xmc4xxx-ebu.yaml [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+  description: |
+    Child device nodes are representing devices connected to the EBU bus.
+
+  properties:
+    reg:
+      type: int
+      required: true
+      description: Selects the region to configure. All the following properties apply to this region.
+
+    device-type:
```
Comment (galak) [inline dts/bindings/memory-controllers/infineon,xmc4xxx-ebu.yaml:145]: please come up with a different property name.  `device-type` is a standard devicetree property.

File: dts/bindings/memory-controllers/infineon,xmc4xxx-ebu.yaml [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+      description: Address bus width configuration.
+      enum:
+        - "16-bit"
+        - "Twin-16-bit"
+        - "32-bit"
+
+    region-enable:
+      type: boolean
+      description: |
+        Enables the address space 0x6x000000-. This address space supports code
```
Comment (galak) [inline dts/bindings/memory-controllers/infineon,xmc4xxx-ebu.yaml:172]: This description doesn't seem finished.

File: dts/bindings/memory-controllers/infineon,xmc4xxx-ebu.yaml [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        execution.
+
+    alternate-region-enable:
+      type: boolean
+      description: Enables the non-executable alternate address space 0xax000000-.
```
Comment (galak) [inline dts/bindings/memory-controllers/infineon,xmc4xxx-ebu.yaml:177]: This description doesn't seem finished.

**Review by galak — CHANGES_REQUESTED** [review]
need to rework the `device-type` property in dts binding.

File: dts/bindings/memory-controllers/infineon,xmc4xxx-ebu.yaml (hunk 1 above)
Comment (talih0) [reply dts/bindings/memory-controllers/infineon,xmc4xxx-ebu.yaml]: I added a infineon prefix to the child bindings

File: dts/bindings/memory-controllers/infineon,xmc4xxx-ebu.yaml (hunk 2 above)
Comment (talih0) [reply dts/bindings/memory-controllers/infineon,xmc4xxx-ebu.yaml]: I added more details in the description.

File: dts/bindings/memory-controllers/infineon,xmc4xxx-ebu.yaml (hunk 3 above)
Comment (talih0) [reply dts/bindings/memory-controllers/infineon,xmc4xxx-ebu.yaml]: I added more details in the description.

**Comment by talih0** [comment]
Reopen PR.
Waiting for https://github.com/zephyrproject-rtos/zephyr/pull/50880 for to get merged before rebasing.

**Comment by talih0** [comment]
Ping pls. 
The memc driver was tested on XMC4800 automation board to interface with sd ram (tests/drivers/memc/ram) and on a custom board to interface with FPGA.

**Comment by talih0** [comment]
I'll rebase this when https://github.com/zephyrproject-rtos/zephyr/pull/55435 gets merged

**Comment by talih0** [comment]
@npal-cy, any chance you could have a look at this PR? 
The main part is the EBU driver (MEMC) for external interfaces. I added XMC4800 automation board to test the EBU driver with external SDRAM.

File: dts/arm/infineon/xmc4xxx.dtsi [hunk 4]
```diff
@@ -51,6 +51,14 @@
 	};
 
 	soc {
+		ebu: ebu@58008000 {
+		    compatible = "infineon,xmc4xxx-ebu";
```
Comment (npal-cy) [inline dts/arm/infineon/xmc4xxx.dtsi:42]: change space to tab

**Comment by talih0** [comment]
@npal-cy thanks, for your comments. I'll update the PR once the xmc47 relax kit is merged.

File: drivers/memc/memc_xmc4xxx.c [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+	bool sdram_enabled = false;
+
+	ret = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	XMC_EBU_Enable(cfg->ebu);
+	cfg->ebu->CLC = cfg->clc;
+	while ((cfg->ebu->CLC & CLC_ACK_Msk) >> 4 != cfg->clc)
```
[the remark this answers is not in the document]
Comment (talih0) [reply drivers/memc/memc_xmc4xxx.c:55]: I don't think a timeout is needed. The same approach is used in XMCLib (xmc_ebu.c, xmc_ebu.h).

File: drivers/memc/memc_xmc4xxx.c [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+	cfg->ebu->USERCON = cfg->usercon;
+
+	for (int i = 0; i < cfg->region_len; i++) {
+		const struct memc_xmc4xxx_region_config *region_config;
+		int region_index;
+
+		region_config = &cfg->region_config[i];
+		region_index = region_config->region_index;
+
+		if (region_index >= 4) {
```
[the remark this answers is not in the document]
Comment (talih0) [reply drivers/memc/memc_xmc4xxx.c:67]: I can't see an XMC define, but I'll add one to the file, thanks

**Review by npal-cy — DISMISSED** [review]

**Review by ifyall — APPROVED** [review]

**Comment by talih0** [comment]
I rebased to get a fresh CI run

**Comment by talih0** [comment]
@npal-cy, could you re-ack when you have a moment, please? The ci errors are from a different board.
It would be great to get this merged before the freeze for the next releasse

**Review by npal-cy — APPROVED** [review]

**Comment by talih0** [comment]
@galak, can you another look at this PR pls?

**Comment by talih0** [comment]
@galak, ping, can you have another look pls

**Pull request closed without merging**